### PR TITLE
[ADD] payment_flutterwave: integrate with "Flutterwave standard" gateway

### PR DESCRIPTION
**Pull request by AntoineVDV**
task-2759117

See also:
- https://github.com/odoo/documentation/pull/1939

### CODE REVIEW

File: addons/payment/tests/common.py [hunk 1]
```diff
@@ -96,6 +96,7 @@ def _get_payment_method_information(self):
         cls.acquirer = cls.dummy_acquirer
         cls.amount = 1111.11
         cls.company = cls.env.company
+        cls.company_id = cls.company.id
```
Comment (hote-odoo) [inline addons/payment/tests/common.py:99]: why this change ?

File: addons/payment/tests/utils.py [hunk 2]
```diff
@@ -40,10 +40,13 @@ def _extract_values_from_html_form(self, html_form):
         :rtype: dict[str:str]
         """
         html_tree = objectify.fromstring(html_form)
-        action = html_tree.get('action')
-        inputs = {form_input.get('name'): form_input.get('value') for form_input in html_tree.input}
+        if hasattr(html_tree, 'input'):
+            inputs = {input_.get('name'): input_.get('value') for input_ in html_tree.input}
+        else:
+            inputs = {}
```
Comment (hote-odoo) [inline addons/payment/tests/utils.py:46]: Same as above, should this not be a separated commit ?

File: addons/payment/tests/common.py (hunk 1 above)
Comment (AntoineVDV) [reply addons/payment/tests/common.py:99]: `env.company` is a property and some IDEs have issues with `id` being read from `self.company` in tests, so I extracted the value from `cls.company` here.

File: addons/payment/tests/utils.py (hunk 2 above)
Comment (AntoineVDV) [reply addons/payment/tests/utils.py:46]: This change was made necessary by a test in `payment_flutterwave` so I think it's fine like this, in the same commit.

File: addons/payment_flutterwave/models/payment_acquirer.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+_logger = logging.getLogger(__name__)
+
+
+class PaymentAcquirer(models.Model):
+    _inherit = 'payment.acquirer'
+
+    provider = fields.Selection(
+        selection_add=[('flutterwave', "Flutterwave")], ondelete={'flutterwave': 'set default'}
+    )
```
Comment (hote-odoo) [inline addons/payment_flutterwave/models/payment_acquirer.py:23]:
```suggestion
provider = fields.Selection(
        selection_add=[('flutterwave', "Flutterwave")], 
        ondelete={'flutterwave': 'set default'}
    )
```

File: addons/payment_flutterwave/models/payment_transaction.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+        :return: None
+        :raise ValidationError: If inconsistent data were received.
+        """
+        super()._process_notification_data(notification_data)
+        if self.provider != 'flutterwave':
+            return
+
+        # Verify the notification data.
+        verification_response_content = self.acquirer_id._flutterwave_make_request(
+            'transactions/verify_by_reference', payload={'tx_ref': self.reference}, method='GET'
```
Comment (chevalierv) [inline addons/payment_flutterwave/models/payment_transaction.py:141]: The payload received in `_flutterwave_make_request()` is given in the body as a JSON object. However, according to Flutterwave documentation, `tx_ref` should be passed as a query parameter. This causes a bad request error at the return of the Flutterwave payment page.

**Review by chevalierv — CHANGES_REQUESTED** [review]

File: addons/payment_flutterwave/controllers/main.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+                )
+                signature = request.httprequest.headers.get('verif-hash')
+                self._verify_notification_signature(signature, tx_sudo)
+
+                # Handle the notification data.
+                notification_data = data['data']
+                tx_sudo._handle_notification_data('flutterwave', notification_data)
+            except ValidationError:  # Acknowledge the notification to avoid getting spammed.
+                _logger.exception("Unable to handle the notification data; skipping to acknowledge")
+        return request.make_json_response('')
```
Comment (chevalierv) [inline addons/payment_flutterwave/controllers/main.py:60]:
```suggestion
        return odoo.http.Response(status=204)  # No Content
```

File: addons/payment_flutterwave/models/payment_acquirer.py [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+        :raise ValidationError: If an HTTP error occurs.
+        """
+        self.ensure_one()
+
+        url = url_join('https://api.flutterwave.com/v3/', endpoint)
+        headers = {'Authorization': f'Bearer {self.flutterwave_secret_key}'}
+        try:
+            if method == 'GET':
+                response = requests.get(url, params=payload, headers=headers, timeout=10)
+            else:
```
Comment (chevalierv) [inline addons/payment_flutterwave/models/payment_acquirer.py:84]: What if `method` equals `PUT` ?

File: addons/payment_flutterwave/__manifest__.py [hunk 7]
```diff
@@ -0,0 +1,20 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+{
+    'name': "Payment Acquirer: Flutterwave",
+    'version': '2.0',
```
[the remark this answers is not in the document]
Comment (AntoineVDV) [reply addons/payment_flutterwave/__manifest__.py:5]: It's like this for all payment acquirers so I didn't feel like breaking the tradition for Flutterwave :)

File: addons/payment_flutterwave/controllers/main.py (hunk 5 above)
Comment (AntoineVDV) [reply addons/payment_flutterwave/controllers/main.py:60]: Yeah that's an idea but I'm not sure if it really adds any value over answering with a plain HTTP 200 that doesn't require a comment?

File: addons/payment_flutterwave/models/payment_acquirer.py (hunk 6 above)
Comment (AntoineVDV) [reply addons/payment_flutterwave/models/payment_acquirer.py:84]: It never equals `PUT` and it's not an API method so no need to overcomplicate for nothing.

**Review by chevalierv — APPROVED** [review]

**Comment by AntoineVDV** [comment]
@robodoo rebase-ff

**Review by odony — APPROVED** [review]
@robodoo override=ci/security
- `requests`: 1 GET and 1 POST with no injectable URL domain and a timeout 👌
- hmac: compare_digest 👌 
- group-restriction for cred fields + password=True

**Comment by AntoineVDV** [comment]
@robodoo r+
